# Step 16. Cohort B: projection and federation

Reads `step12_panels.rds`. Writes `step16_projection.rds`.

Cohort B has one surviving module, so its column axis is a single block. Two ways to give B the
structure the other cohorts found, and they are not the same thing:

- Projection carries a module *definition* into B. Loadings come from the source cohort and are
  never recomputed; the new samples are scaled with the source's per-protein mean and sd, not
  their own. Re-centring on B would move B's mean to zero, which is precisely the difference being
  measured. The round-trip gate must return r = 1.0000 or the code is refitting, not projecting.
- Federation pools sufficient statistics so no site sees another's patients. It is governed by
  the release rule `p < n`, because `rank(G) = min(n, p)` and a Gram matrix from a panel wider than
  the cohort can be inverted back toward individual rows.

In [1]:
source("../src/paths.R")
suppressMessages(library(WGCNA))
options(stringsAsFactors = FALSE); set.seed(SEED)
P <- readRDS(art("step12_panels.rds"))
D <- P$D; spec <- P$spec; SITES <- P$SITES
W  <- lapply(SITES, function(s) readRDS(art("wgcna_%s.rds", s))); names(W) <- SITES

build_traits <- function(m, spec) {
  out <- data.frame(row.names = rownames(m))
  for (i in seq_len(nrow(spec))) {
    v <- m[[spec$source_column[i]]]
    out[[spec$name[i]]] <- switch(spec$type[i],
      numeric = as.numeric(as.character(v)),
      binary  = as.numeric(v == "Positive"),
      ordinal = { lvl <- unique(v[!is.na(v) & v != ""])
                  lvl <- lvl[order(as.numeric(sub("-.*", "", lvl)))]
                  as.integer(factor(v, levels = lvl, ordered = TRUE)) })
  }
  out
}
TR <- lapply(SITES, function(s) build_traits(W[[s]]$meta, spec)); names(TR) <- SITES
N_MIN <- min(sapply(W, function(w) nrow(w$X)))

suppressMessages(library(WGCNA)); options(stringsAsFactors=FALSE); set.seed(SEED)

# ---- projection: loadings from the SOURCE only, never recomputed ---------
project_module <- function(md0, X, mods, NEW){
  g  <- colnames(X)[mods==md0]
  mu <- colMeans(X[,g,drop=FALSE]); sdv <- apply(X[,g,drop=FALSE],2,sd)
  keep <- sdv>0; g<-g[keep]; mu<-mu[keep]; sdv<-sdv[keep]
  Zs <- scale(X[,g,drop=FALSE], center=mu, scale=sdv)
  v  <- svd(Zs, nu=0, nv=1)$v[,1]
  e  <- as.vector(Zs %*% v)
  if (cor(e, rowMeans(Zs)) < 0){ v <- -v; e <- -e }
  Zn <- scale(as.matrix(NEW[,g,drop=FALSE]), center=mu, scale=sdv)
  list(sle=e, new=as.vector(Zn %*% v))
}

cat("=== ROUTE 1: PROJECTION -- B scored on A's and C's module definitions ===\n")
proj <- list()
for (src in c("A","C")){
  d <- D[[paste(src,"all15")]]
  # A cohort can reach this point with nothing to project. Step 12 drops any module
  # covering 1% or more of the panel, and for a cohort whose only trait-associated
  # module is one of those, `keep` is empty. That is a result about the cohort, so it
  # is reported here rather than being allowed to fail as min() on an empty set.
  if (!length(d$keep)) {
    cat(sprintf("\n  %s -> B : no module survives step 12's 1%% size exclusion, so %s has\n            nothing to project. See step 12 for what was excluded and why.\n", src, src))
    next
  }
  Xs <- W[[src]]$X; mods <- W[[src]]$mods
  Xb <- W$B$X
  ME <- moduleEigengenes(Xs, mods)$eigengenes
  Pm <- lapply(d$keep, function(k) project_module(k, Xs, mods, Xb)); names(Pm) <- d$keep
  gate <- sapply(d$keep, function(k) abs(cor(Pm[[k]]$sle, ME[[paste0("ME",k)]])))
  cat(sprintf("\n  %s -> B : %d modules, round-trip r min %.4f median %.4f\n",
              src, length(d$keep), min(gate), median(gate)))
  stopifnot(min(gate) > 0.99)
  cat("  GATE PASSED -- projection, not a refit\n")
  trB <- TR$B[rownames(Xb), , drop=FALSE]
  sc  <- sapply(d$keep, function(k) Pm[[k]]$new); rownames(sc) <- rownames(Xb)
  # Spearman and the per-pair n, as in step 04, which explains both.
  r   <- cor(sc, trB, method="spearman", use="pairwise.complete.obs")
  nn  <- crossprod(!is.na(sc), !is.na(as.matrix(trB)))
  p   <- 2*pt(-abs(r*sqrt((nn-2)/(1-r^2))), nn-2)
  q   <- matrix(p.adjust(p,"BH"), nrow=nrow(p), dimnames=dimnames(p))
  h   <- which(q<0.05, arr.ind=TRUE)
  cat(sprintf("  %s's modules tested against B's traits: %d of %d at FDR 5%%\n",
              src, nrow(h), length(q)))
  if (nrow(h)) print(data.frame(module=rownames(r)[h[,1]], trait=colnames(r)[h[,2]],
      r=round(r[h],2), q=signif(q[h],2), n_prot=sapply(rownames(r)[h[,1]],
      function(k) sum(mods==k)))[order(-abs(r[h])),], row.names=FALSE)
  proj[[src]] <- list(scores=sc, gate=gate, r=r, q=q)
}

cat("\n\n=== ROUTE 2: FEDERATION -- release rule p < n ===\n")
cat(sprintf("  smallest cohort n = %d, so a releasable panel needs p <= %d proteins\n\n", N_MIN, N_MIN-1))
fed <- do.call(rbind, lapply(names(D), function(nm){ d<-D[[nm]]
  data.frame(cohort=d$cohort, cond=d$cond, probes=length(d$sel),
             proteins=n_proteins(d$sel), n=N_MIN,
             releasable=n_proteins(d$sel) < N_MIN) }))
print(fed, row.names=FALSE)
cat(sprintf("\n  panels that pass: %d of %d. Federation of these panels is NOT possible.\n",
            sum(fed$releasable), nrow(fed)))
cat("  (step 10 federates an 80-protein panel exactly; none of these is that small.)\n")

cat("\n\n=== DIAGNOSTIC: why is A the most discriminating? ===\n")
cat("\nmodules and associations per cohort (all15):\n")
print(do.call(rbind, lapply(SITES, function(s){ d<-D[[paste(s,"all15")]]
  m <- W[[s]]$mods
  data.frame(cohort=s, n=nrow(W[[s]]$X), modules=length(setdiff(unique(m),"grey")),
    grey=sum(m=="grey"), largest=max(table(m[m!="grey"])),
    largest_pct=round(100*max(table(m[m!="grey"]))/7288,1),
    associated=length(d$sig)) })), row.names=FALSE)

cat("\ntrait spread per cohort (sd for numeric, positive count for binary):\n")
sp <- do.call(rbind, lapply(spec$name, function(t){
  row <- data.frame(trait=t, type=spec$type[spec$name==t])
  for (s in SITES){ v <- TR[[s]][[t]]
    row[[s]] <- if (spec$type[spec$name==t]=="binary") sum(v==1,na.rm=TRUE) else round(sd(v,na.rm=TRUE),2) }
  row }))
print(sp, row.names=FALSE)

cat("\nbatch composition (the split was stratified on Batch only):\n")
print(do.call(rbind, lapply(SITES, function(s)
  data.frame(cohort=s, t(as.matrix(table(W[[s]]$meta$Batch)))))), row.names=FALSE)

saveRDS(list(proj = proj, fed = fed), art("step16_projection.rds"))


=== ROUTE 1: PROJECTION -- B scored on A's and C's module definitions ===



  A -> B : 5 modules, round-trip r min 1.0000 median 1.0000
  GATE PASSED -- projection, not a refit
  A's modules tested against B's traits: 0 of 75 at FDR 5%

  C -> B : 1 modules, round-trip r min 1.0000 median 1.0000
  GATE PASSED -- projection, not a refit
  C's modules tested against B's traits: 0 of 15 at FDR 5%




=== ROUTE 2: FEDERATION -- release rule p < n ===


  smallest cohort n = 86, so a releasable panel needs p <= 85 proteins



 cohort   cond probes proteins  n releasable
      A  all15    166      153 86      FALSE
      A varsel     93       83 86       TRUE
      A  union    134      121 86      FALSE
      B  all15     26       22 86       TRUE
      B varsel      0        0 86       TRUE
      B  union     26       22 86       TRUE
      C  all15     38       38 86       TRUE
      C varsel    261      260 86      FALSE
      C  union    108      108 86      FALSE



  panels that pass: 5 of 9. Federation of these panels is NOT possible.


  (step 10 federates an 80-protein panel exactly; none of these is that small.)




=== DIAGNOSTIC: why is A the most discriminating? ===



modules and associations per cohort (all15):


 cohort  n modules grey largest largest_pct associated
      A 87      33 1107    2847        39.1          8
      B 87      15 1658    4522        62.0          3
      C 86      37 1376    3409        46.8          5



trait spread per cohort (sd for numeric, positive count for binary):


            trait    type      A      B      C
        SLEDAI_2K numeric   3.55   2.94   2.78
         C3_level numeric   0.30   0.32   0.29
         C4_level numeric   0.11   0.12   0.11
   Duration_years numeric  10.30  10.95   9.70
 Lymphocyte_count numeric   0.85   0.90   0.54
             uPCR numeric 279.20 151.51 179.89
       Creatinine numeric  44.25  60.42  73.78
        Sm_status  binary  21.00  18.00  24.00
    RNP_68_status  binary  11.00   7.00  18.00
     RNP_A_status  binary  20.00  20.00  28.00
     Ro_52_status  binary  24.00  30.00  25.00
     Ro_60_status  binary  36.00  46.00  38.00
        La_status  binary  11.00   9.00  11.00
     dsDNA_status  binary  26.00  28.00  16.00
         Age_band ordinal   2.89   2.71   2.67



batch composition (the split was stratified on Batch only):


 cohort  A  B
      A 69 18
      B 69 18
      C 69 17


## Findings

**Projection is a valid projection, and it finds nothing at B.** The gate passes at r = 1.0000 both
ways, so these are A's and C's loadings applied to B's patients and not a refit. But A→B gives **0 of
75** module x trait associations at FDR 5%, and C→B gives **0 of 15**. Scoring B's patients on another
cohort's whole panel, then testing, recovers nothing.

That is a negative result and it belongs here. Step 18 does the same thing one module at a time and
does find signal at B — the interferon module reaches 10 associations there. The difference is what is
being corrected over: 75 tests across five whole modules against 15 traits, versus 15 tests for one
module. A panel-wide projection spends its multiple-testing budget on modules that carry nothing.

**Most of these panels cannot be federated whole.** Of the 9 cohort x condition panels, 4 clear
`p < n = 86`: A `varsel` at 83 proteins, B `all15` and `union` at 22, C `all15` at 38. The rest are
121 to 260 proteins. Step 18 releases modules individually instead, where every one of the seven
clears the rule at 22 to 39 proteins.

## Where this was run

Rendered notebooks are committed, so each records the machine, the R and the
package versions that produced its output.


In [2]:
# Provenance. The rendered notebook is committed, so it records where it ran.
# The cell is identical in every notebook and is explained in step 00.
si <- Sys.info()
cat("run on      :", si[["nodename"]], "(", si[["sysname"]], si[["release"]], ")\n")
cat("date        :", format(Sys.time(), "%Y-%m-%d %H:%M %Z"), "\n")
cat("R           :", R.version.string, "|", R.version$platform, "\n")
cat("R comes from:", R.home(), "\n")
cat("packages    :\n")
for (p in c("WGCNA", "ComplexHeatmap", "circlize", "ReactomePA", "reactome.db",
            "org.Hs.eg.db", "sva", "VarSelLCM"))
  cat(sprintf("   %-16s %s\n", p,
              tryCatch(as.character(packageVersion(p)), error = function(e) "not installed")))

run on      : Annes-MacBook-Pro-193.local ( Darwin 27.0.0 )


date        : 2026-09-28 01:04 EDT 


R           : R version 4.4.3 (2025-02-28) | x86_64-apple-darwin13.4.0 


R comes from: /Users/adeslatt/miniforge3/envs/endotypes-proteomics/lib/R 


packages    :


   WGCNA            1.74
   ComplexHeatmap   2.22.0
   circlize         0.4.18
   ReactomePA       1.50.0
   reactome.db      1.89.0
   org.Hs.eg.db     3.20.0
   sva              3.54.0
   VarSelLCM        2.1.3.2
